### ISSUE TITLE
Tabnet classifier training Reproducibility on different machines

### ISSUE THREAD

**Issue by Ananthasireesh** [dreamquark-ai/tabnet#379]
abc

**Comment by Ananthasireesh** [comment]
code for running it on different machines:


import torch
import numpy as np
import pandas as pd
import os
from feature_engine.encoding import PRatioEncoder
from collections import defaultdict
from sklearn.preprocessing import LabelEncoder
from pytorch_tabnet.tab_model import TabNetClassifier, TabNetRegressor
import warnings
import sys
if not sys.warnoptions:
    warnings.simplefilter("ignore")

data_df = pd.read_csv('https://storage.googleapis.com/tf-datasets/titanic/train.csv')
data_df.head(2)

#this code will get train/test/val indexes, impute null, extract target variable.

#split percentage
train_set = 80
validation_set = 10
test_set = 10

np.random.seed(666) # test -train-split 
df_filtered = data_df
df_filtered.reset_index(drop=True, inplace=True)

if "Set" not in df_filtered.columns:
    df_filtered["Set"] = np.random.choice(["train", "valid", "test"],
                                            p=[round(train_set / 100, 2), round(validation_set / 100, 2),
                                            round(test_set / 100, 2)], size=(df_filtered.shape[0],))

print("\nInput Data Samples: ", len(df_filtered))
train_indices = df_filtered[df_filtered.Set == "train"].index
valid_indices = df_filtered[df_filtered.Set == "valid"].index
test_indices = df_filtered[df_filtered.Set == "test"].index

df_test = df_filtered[df_filtered["Set"] == 'test']
df_valid = df_filtered[df_filtered["Set"] == 'valid']

print("Training Data Samples: ", len(train_indices))
print("Validation Data Samples: ", len(valid_indices))
print("Test Data Samples: ", len(test_indices))

features = list(df_filtered.columns[1:-1])
print("Target Variable: ", df_filtered.columns[0])
print("Training Features: ", ', '.join(features))

target_labels = df_filtered.iloc[:, 0].unique()

try:
    print("Training Labels: ", ', '.join(target_labels))
except:
    print("Training Labels: ", target_labels)

# Appending the target variable
categorical_columns.append(df_filtered.columns[0])

d = defaultdict(LabelEncoder)
categorical_dims = {}

if len(categorical_columns) > 1:
    for col in categorical_columns[0:-1]:
        if df_filtered[col].dtype == 'float64':
            df_filtered[col] = df_filtered[col].fillna(0).astype(np.int64, errors='ignore')
        df_filtered[col] = df_filtered[col].astype('str')

if len(categorical_columns) > 1:
    cat_col_typelist = []
    for catColType in categorical_columns[0:-1]:
        if df_filtered[catColType].dtype not in ['int32', 'int64', 'datetime64','datetime64[ns]']:
            cat_col_typelist.append(catColType)

for col in categorical_columns:
    df_filtered[col] = df_filtered[col].fillna("IDoNotKnow")
    df_filtered[col] = d[col].fit_transform(df_filtered[col].values)
    categorical_dims[col] = len(d[col].classes_)

for col in features:
    if col not in categorical_columns:
        df_filtered[col] = pd.to_numeric(df_filtered[col], errors='coerce')
        df_filtered[col] = df_filtered[col].fillna(0)
    else:
        pass

cat_idxs = [i for i, f in enumerate(features) if f in categorical_columns]
cat_dims = [categorical_dims[f] for i, f in enumerate(features) if f in categorical_columns]


#data splitting using the indexes created
X_train = df_filtered[features].values[train_indices]
y_train = df_filtered.iloc[:, 0].values[train_indices]
X_valid = df_filtered[features].values[valid_indices]
y_valid = df_filtered.iloc[:, 0].values[valid_indices]
X_test = df_filtered[features].values[test_indices]
y_test = df_filtered.iloc[:, 0].values[test_indices]


max_epochs = 10

learning_rate = 0.02
batchsize = 31
virtual_batch_size = 4
n_independent = 2
n_shared = 2
n_d = 8
lambda_sparse = 0.001
patience = 10

clf = TabNetClassifier(cat_idxs=cat_idxs,
                           cat_dims=cat_dims,
                           cat_emb_dim=1,
                           optimizer_fn=torch.optim.Adam,
                           optimizer_params=dict(lr=learning_rate),
                           scheduler_params={"step_size": 50,  # how to use learning rate scheduler
                                             "gamma": 0.9},
                           scheduler_fn=torch.optim.lr_scheduler.StepLR,
                           mask_type='entmax',  # "sparsemax"
                           n_independent=n_independent,
                           n_shared=n_shared,
                           n_d=n_d,
                           lambda_sparse=lambda_sparse,
                           seed=10
                           )

print("\nTraining...")
clf.fit(
    X_train=X_train, y_train=y_train, eval_set=[(X_valid, y_valid)],
    max_epochs=max_epochs, patience=patience,
    batch_size=batchsize, virtual_batch_size=virtual_batch_size,
    num_workers=0,
    weights=1,
    drop_last=False,
    eval_metric=['balanced_accuracy']
)

**Comment by Optimox** [comment]
Hello @Ananthasireesh,

Are you sure that you are running the exact same code with the exact same version of pytorch-tabnet ?

Can you try to call this function with same seed at the beginning of your code ?

```
def seed_everything(seed):
    """
    Seeds basic parameters for reproductibility of results.

    Args:
        seed (int): Number of the seed.
    """
    random.seed(seed)
    os.environ["PYTHONHASHSEED"] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False
```

**Comment by Ananthasireesh** [comment]
Hello @Optimox , Thanks for your quick revert. 

we are using same version of pytorch-tabnet  that is 3.1.1, and we are also using exact same code 

We did try this code yesterday and it starting and it did not work and we even tried now to reconfirm, still no luck of getting results. 

here am pasting both results on mine and friends system, after trying the code provided by you at start of the code, i can also post my code if needed. Thanks in advance

Results on My machine:

Device used : cpu

Training...
epoch 0  | loss: 0.74028 | val_0_balanced_accuracy: 0.56816 |  0:00:00s
epoch 1  | loss: 0.68302 | val_0_balanced_accuracy: 0.56463 |  0:00:01s
epoch 2  | loss: 0.7286  | val_0_balanced_accuracy: 0.59694 |  0:00:01s
epoch 3  | loss: 0.67813 | val_0_balanced_accuracy: 0.63925 |  0:00:02s
epoch 4  | loss: 0.68109 | val_0_balanced_accuracy: 0.7074  |  0:00:02s
epoch 5  | loss: 0.71609 | val_0_balanced_accuracy: 0.75029 |  0:00:03s
epoch 6  | loss: 0.67044 | val_0_balanced_accuracy: 0.59166 |  0:00:03s
epoch 7  | loss: 0.65305 | val_0_balanced_accuracy: 0.6181  |  0:00:04s
epoch 8  | loss: 0.63927 | val_0_balanced_accuracy: 0.65041 |  0:00:04s
epoch 9  | loss: 0.64167 | val_0_balanced_accuracy: 0.65864 |  0:00:05s
Stop training because you reached max_epochs = 10 with best_epoch = 5 and best_val_0_balanced_accuracy = 0.75029
Best weights from best epoch are automatically used!


Results on my friends machine:
Device used : cpu

Training...
epoch 0  | loss: 0.78176 | val_0_balanced_accuracy: 0.66686 |  0:00:01s
epoch 1  | loss: 0.69598 | val_0_balanced_accuracy: 0.43478 |  0:00:02s
epoch 2  | loss: 0.69559 | val_0_balanced_accuracy: 0.51469 |  0:00:03s
epoch 3  | loss: 0.6712  | val_0_balanced_accuracy: 0.64806 |  0:00:04s
epoch 4  | loss: 0.6735  | val_0_balanced_accuracy: 0.61163 |  0:00:05s
epoch 5  | loss: 0.6807  | val_0_balanced_accuracy: 0.59577 |  0:00:06s
epoch 6  | loss: 0.65395 | val_0_balanced_accuracy: 0.71269 |  0:00:06s
epoch 7  | loss: 0.68537 | val_0_balanced_accuracy: 0.59107 |  0:00:07s
epoch 8  | loss: 0.65966 | val_0_balanced_accuracy: 0.5047  |  0:00:08s
epoch 9  | loss: 0.65293 | val_0_balanced_accuracy: 0.59636 |  0:00:09s
Stop training because you reached max_epochs = 10 with best_epoch = 6 and best_val_0_balanced_accuracy = 0.71269
Best weights from best epoch are automatically used!

**Comment by Optimox** [comment]
To be honest I don't know why this is happening...

I doubt that I'll be able to help : https://discuss.pytorch.org/t/reproducibility-over-different-machines/63047

But why do you need to reproduce the exact same training?
It's just a random seed issue, you should probably just increase number of epochs and reduce your learning rate until convergence.

**Comment by Ananthasireesh** [comment]
Thanks @Optimox , need of the day is that we are trying to create a semi ML flask application that people can use where they just has to put their input file, give categorical variables and then a model can be trained, and that model can be utilized by people to infer the results. Hope you got me. :)

we will try increasing number of epochs and reduce learning rate until convergence, Thanks for your time.
